File: src/sources/fixture.ts
import { readFile } from "node:fs/promises";
import type { CityConfig, DateWindow, Evidence, Listing } from "../models.js";
import { estimateTransitAccessibility } from "../transit.js";
import type { SourceAdapter } from "./base.js";

interface FixtureAmenity {
  status: Evidence["status"];
  detail: string;
}

interface FixtureListing {
  source_listing_id: string;
  name: string;
  url: string;
  city: string;
  neighborhood?: string;
  location_text?: string;
  latitude?: number;
  longitude?: number;
  room_type?: string;
  nightly_price_eur: number;
  available_stay_lengths: number[];
  rating?: number;
  rating_scale?: number;
  review_count?: number;
  cancellation_policy?: string;
  source_reliability?: number;
  amenities?: Record<string, FixtureAmenity>;
  [key: string]: unknown;
}

interface FixtureFile {
  listings: FixtureListing[];
}

export class FixtureSource implements SourceAdapter {
  readonly name = "fixture";

  constructor(private readonly path: string) {}

  async search(cities: CityConfig[], windows: DateWindow[]): Promise<Listing[]> {
    const data = JSON.parse(await readFile(this.path, "utf8")) as FixtureFile;
    const cityByName = new Map(cities.map((city) => [city.name.toLowerCase(), city]));
    const listings: Listing[] = [];

    for (const item of data.listings) {
      const city = cityByName.get(item.city.toLowerCase());
      if (!city) continue;
      const window = windows.find((candidate) => item.available_stay_lengths.includes(candidate.nights));
      if (!window) continue;
      const listing = this.listingFromItem(item, window);
      listing.transit = estimateTransitAccessibility(listing, city);
      listings.push(listing);
    }

    return listings;
  }

  private listingFromItem(item: FixtureListing, window: DateWindow): Listing {
    const amenities: Record<string, Evidence> = {};
    for (const [key, value] of Object.entries(item.amenities ?? {})) {
      amenities[key] = { status: value.status, detail: value.detail, source: this.name };
    }
    const nightly = Number(item.nightly_price_eur);
    return {
      source: this.name,
      sourceReliability: Number(item.source_reliability ?? 0.6),
      sourceListingId: String(item.source_listing_id),
      name: item.name,
      url: item.url,
      city: item.city,
      neighborhood: item.neighborhood ?? "",
      dates: window,
      totalPriceEur: round(nightly * window.nights, 2),
      nightlyPriceEur: nightly,
      rating: item.rating ?? null,
      ratingScale: Number(item.rating_scale ?? 10),
      reviewCount: item.review_count ?? null,
      locationText: item.location_text ?? "",
      latitude: item.latitude ?? null,
      longitude: item.longitude ?? null,
      roomType: item.room_type ?? "",
      cancellationPolicy: item.cancellation_policy ?? "",
      amenities,
      transit: { status: "missing", detail: "No transit evidence." },
      raw: item
    };
  }
}

function round(value: number, digits: number): number {
  return Number(value.toFixed(digits));
}

File: travel_deals/sources/fixture.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from travel_deals.models import CityConfig, DateWindow, Evidence, Listing
from travel_deals.sources.base import SourceAdapter
from travel_deals.transit import estimate_transit_accessibility


class FixtureSource(SourceAdapter):
    name = "fixture"

    def __init__(self, path: Path):
        self.path = path

    def search(self, cities: list[CityConfig], windows: list[DateWindow]) -> list[Listing]:
        city_by_name = {city.name.lower(): city for city in cities}
        data = json.loads(self.path.read_text(encoding="utf-8"))
        listings: list[Listing] = []
        for item in data["listings"]:
            city = city_by_name.get(item["city"].lower())
            if city is None:
                continue
            for window in windows:
                if window.nights not in item.get("available_stay_lengths", []):
                    continue
                nightly = float(item["nightly_price_eur"])
                listing = self._listing_from_item(item, window, nightly)
                listing.transit = estimate_transit_accessibility(listing, city)
                listings.append(listing)
                break
        return listings

    def _listing_from_item(self, item: dict[str, Any], window: DateWindow, nightly: float) -> Listing:
        amenities = {
            key: Evidence(value["status"], value["detail"], self.name)
            for key, value in item.get("amenities", {}).items()
        }
        return Listing(
            source=self.name,
            source_reliability=float(item.get("source_reliability", 0.6)),
            source_listing_id=str(item["source_listing_id"]),
            name=item["name"],
            url=item["url"],
            city=item["city"],
            neighborhood=item.get("neighborhood", ""),
            dates=window,
            total_price_eur=round(nightly * window.nights, 2),
            nightly_price_eur=nightly,
            rating=item.get("rating"),
            rating_scale=float(item.get("rating_scale", 10)),
            review_count=item.get("review_count"),
            location_text=item.get("location_text", ""),
            latitude=item.get("latitude"),
            longitude=item.get("longitude"),
            room_type=item.get("room_type", ""),
            cancellation_policy=item.get("cancellation_policy", ""),
            amenities=amenities,
            raw=item,
        )
